tool/src/org/antlr/v4/codegen/target/Python3Target.java

```diff
@@ -87,4 +87,7 @@ public boolean wantsBaseVisitor() {
 	public boolean supportsOverloadedMethods() {
 		return false;
 	}
+
+	@Override
+	public boolean supportsConstants() { return false; }
```

Comment:
Hang on. I'm just not getting this.  Why not just always do this if it's better?  All targets can handle bitsets right?

Reply:
Also, this is turned off for Python but Python is what we want to fix.  Maybe flip the boolean sense of the function name so True means do this new stuff.

Reply:
Yeah, I asked this via email (which got lost), so again:

Hi, silly question - why conditionally fold, why not just always fold for all targets? Seems odd, am I misunderstanding something?

(worth having a 'do not fold' option for debugging purposes, so you can examine the output, but only for human use).

cheers

jan

Reply:
Further, as I can't find the right post and my other email also disappeared, there's a question about the nomenclature of 'folding' and what a 'constant' is:

Ivan says > "is being folded to constant 0xFFFF. But Python doesn't support constant propagation since Python is unable to determine whether a variable is constant or not."

My take is folding = evaluation and a constant here is more accurately described as a literal. 

I love the smell of bikeshedding in the morning...

Reply:
Constant folding is executed for all runtimes during generation. It's used for folding long expressions with bitwise operations. Expressions like this:

```python
if not(((((_la - 1)) & ~0x3f) == 0 and ((1 << (_la - 1)) & ((1 << (PParser.T0 - 1)) | (1 << (PParser.T1 - 1)) | (1 << (PParser.T2 - 1)) | (1 << (PParser.T3 - 1)) | (1 << (PParser.T4 - 1)) | (1 << (PParser.T5 - 1)) | (1 << (PParser.T6 - 1)) | (1 << (PParser.T7 - 1)) | (1 << (PParser.T8 - 1)) | (1 << (PParser.T9 - 1)) | (1 << (PParser.T10 - 1)) | (1 << (PParser.T11 - 1)) | (1 << (PParser.T12 - 1)) | (1 << (PParser.T13 - 1)) | (1 << (PParser.T14 - 1)) | (1 << (PParser.T15 - 1)) | (1 << (PParser.T16 - 1)) | (1 << (PParser.T17 - 1))
```

Constant propagation is used only for targets that don't support constant concept (Python, JavaScript), it differs from folding. For example, in the following code:

```
self.enterRule(localctx, 0, self.RULE_r)
```

`RULE_r` will be replaced with constant literal:

```
self.enterRule(localctx, 0, 0)
```

but it will be saved for compilable languages. I thought it would be more clear without affecting performance, but we can use constant propagation for all targets as well (together with constant folding).

Reply:
You are right about the clarity. Perhaps emit the un-optimised/un-folded code as a comment next to it? I dunno, you are more familiar with the outputs and what's required/nice/

Reply:
Ah! Now I'm starting to understand your terminology. There are TWO things going on: folding and replacing symbols with literals.  

1. We should always fold.
2. When would we propagate ever?  Folding should fix all those issues right and remove refs to constant symbols

Reply:
Propagation and folding are always used for bitwise operations for all targets. But in some other places, there are only refs to constants:

```
def r(self):
    localctx = TestParser.RContext(self, self._ctx, self.state)
    self.enterRule(localctx, 0, 0)
    try:
        self.enterOuterAlt(localctx, 1)
        self.state = 2
        self.match(1)
    except RecognitionException as re:
        localctx.exception = re
        self._errHandler.reportError(self, re)
        self._errHandler.recover(self, re)
    finally:
        self.exitRule()
    return localctx
```

New method `supportConstantPropagation` only affects such places but not bitwise operations.

Reply:
Hmm...In this case I think it's better to see `TestParser.T__0` rather than say 42.  Looking for token matches is how I orient in the code.

In the end, then, I don't think we want the Propagation option.

Reply:
Sorry, I forgot about an important case where constant propagation may affect performance a lot. The following code:

```python
i in [0, 1, 2, 3,
                 4, 5, 6, 7,
                 8, 9, 10, 11,
                 12, 13, 14, 15,
                 16, 17, 18, 19,
                 20, 21, 22, 23,
                 24, 25, 26, 27,
                 28, 29, 30, 31]
```

works up to 10x times faster than the following:

```python
i in [C.C0, C.C1, C.C2, C.C3,
                 C.C4, C.C5, C.C6, C.C7,
                 C.C8, C.C9, C.C10, C.C11,
                 C.C12, C.C13, C.C14, C.C15,
                 C.C16, C.C17, C.C18, C.C19,
                 C.C20, C.C21, C.C22, C.C23,
                 C.C24, C.C25, C.C26, C.C27,
                 C.C28, C.C29, C.C30, C.C31]
```

Result:

```
check_by_range_with_literals_test: 129 ns
check_by_range_test: 1033 ns
```

Also, it generates smaller bytecode. You can check it using the script [check-range-vs-mask](https://gist.github.com/KvanTTT/e3b355f7e321fe7f52e11ea1aa0ecbce#file-check-range-vs-mask-py)

Such code is used in a lot of places where other languages use `switch` statement.

We can resolve the problem in two ways:

1. Use constant propagation as I suggested. No changes are needed.
2. Use bitwise checking as @amykyta3 suggested. I.e. use `(1 << token) & 0x20000000f00000003ffffff000000000 != 0` instead of `i in [0, 1, 2, 3]`. It seems like it works a bit slower for most cases.

What way do you prefer?

Reply:
Very interesting. Yes I can imagine in python that would be much much slower because it is doing symbol look up. Is there a way to distinguish between singletons like `match(C.TYPEDEF)` and sets as you have above? In other words how about we do your constant propagation but we limited to sets. Once you go beyond one token it's unlikely people are trying to read that for debugging purposes.

Reply:
> Is there a way to distinguish between singletons like match(C.TYPEDEF) and sets as you have above?

Yes, it's possible, but I implemented constant propagation for such cases because they are located in parser code and may affect the performance in theory as well. I can do it only for sets if you wish.

Reply:
yeah, readability is important and match(T) should be left as-is. thanks!

Reply:
Done. I managed to fix it without introducing the new member `supportsConstant` or similar.

Reply:
Some tests are failing, I'll take a look.